python/ray/data/dataset.py

```diff
@@ -317,88 +318,124 @@ def map_batches(
         fn_constructor_kwargs: Optional[Dict[str, Any]] = None,
         **ray_remote_args,
     ) -> "Dataset[Any]":
-        """Apply the given function to batches of records of this dataset.
+        """Apply the given function to batches of data.
 
-        The format of the data batch provided to ``fn`` can be controlled via the
-        ``batch_format`` argument, and the output of the UDF can be any batch type.
+        Batches are represented as dataframes, arrays, or lists. The default batch
+        type is determined by your dataset's schema. To determine the default batch
+        type, call :meth:`~Dataset.native_batch_format`. Alternatively, set the batch
+        type with ``batch_format``.
 
-        This is a blocking operation.
+        To learn more about writing functions for :meth:`~Dataset.map_batches`, read
+        :ref:`writing user-defined functions <transform_datasets_writing_udfs>`.
+
+        .. note::
+            This is a blocking operation. The time complexity is :math:`O(n/p)`, where
+            :math:`n` is the dataset size and :math:`p` is the parallelism.
+
+        .. tip::
+            Consider using :class:`~ray.data.preprocessors.BatchMapper`. It's more
+            performant and easier to use.
 
         Examples:
+
+            >>> import pandas as pd
             >>> import ray
-            >>> # Transform python objects.
-            >>> ds = ray.data.range(1000)
-            >>> # Transform batches in parallel.
-            >>> ds.map_batches(lambda batch: [v * 2 for v in batch])
-            Dataset(num_blocks=..., num_rows=1000, schema=<class 'int'>)
-            >>> # Define a callable class that persists state across
-            >>> # function invocations for efficiency.
-            >>> init_model = ... # doctest: +SKIP
+            >>> df = pd.DataFrame({
+            ...     "name": ["?", "Grace Hopper", "Rafael Nadal"],
+            ...     "age": ["22", "85", "36"]
+            ... })
+            >>> ds = ray.data.from_pandas(df)
+            >>> ds
+            Dataset(num_blocks=1, num_rows=3, schema={name: object, age: object})
+
+            Call :meth:`~Dataset.native_batch_format` to determine the default batch
+            type.
+
+            >>> ds.native_batch_format()
+            <class 'pandas.core.frame.DataFrame'>
+
+            Once you know the batch type, define a function that transforms batches
+            of data. :meth:`~Dataset.map_batches` applies the function in parallel.
+
+            >>> def map_fn(batch: pd.DataFrame) -> pd.DataFrame:
+            ...     batch["age"] = batch["age"].astype(int)
```

Comment:
Might want to do something more interesting than a cast for the example UDF?

Reply:
Do you have any suggestions? I tried thinking of something that's both simple and concise (like, few lines of code for both the dataframe and `map_fn`).

Reply:
Hmm ideally it would be something like generating a new column via a simple linear combination of one or more other columns, e.g.
```python
def map_fn(batch: pd.DataFrame) -> pd.DataFrame:
    batch["age_in_dog_years"] = 7 * batch["age"]
    return batch
```

Reply:
I also have ~2 users unsure about whether the change is gonna be on the same column or different.

They were staring at `ds.map_batches(lambda batch: [v * 2 for v in batch])`.

Reply:
@clarkzinzow updated example with your recommendation.

> I also have ~2 users unsure about whether the change is gonna be on the same column or different.

@xwjiang2010 could you elaborate? I'm not sure if I understand.

Reply:
I guess they just didn't realize that it's up to them (the way how they write the udf) to decide whether the new value is applied to the old column or new one. 

So in this case, they can write it in a way that "age" becomes 7 times the original value "in place" or just add a new column with the new age value. 

@clarkzinzow btw, I do that maths all the time with my cats as well!! :)